Redirect navigation flow for 0 value invocies. Fixes #991

emiljoha opened the pull request:
- [x] Functionality in desktop. (Tested on Linux) 
- [x] Functionality in mobile (Tested on android)
- [x] Fix: Background on new view does not properly match views before and after. 
- [x] Make sure text content and placement is consitent with rest of the app.
- [x] Fix: Input field on Desktop becomes 0500 when trying to write 500.
- [x] Error handling. Make sure all error scenarios are considered and handled properly. (Note: My experience in UI coding is limited but I am currently convinced this new feature will handle errors gracefully)

![more-or-less-polished](https://user-images.githubusercontent.com/12907343/73122639-6d977a00-3f87-11ea-8a00-33ca6538b305.gif)

emiljoha committed:
Redirect navigation flow for 0 value invocies.

Detecting a zero amount invoice in checkType we redirect to a seperate view.
Currently only title is changed in the view.

emiljoha committed:
Unit tests for goPayLightningSupplyAmount and checkType.

emiljoha committed:
Unpolished but working desktop zero value payments.

emiljoha committed:
Unpolished zero value invoice support for mobile.

emiljoha commented:
Basic idea of the UX.
![zero-amount-invoice-ux](https://user-images.githubusercontent.com/12907343/72670161-932dfc00-3a3a-11ea-9621-31ab1f3cba0b.gif)

emiljoha commented:
Done

[x] Functionality in desktop. (Tested on Linux) 

[x] Functionality in mobile (Tested on android)

Needs to be fixed: 

[ ] Background on new view does not properly matchview before and after. 

[ ] Error handling. Make sure all error scenarios are considered and handled properly. 

[ ] (Optional) Receiving zero value invoices give notice of receiving 0 sats. Optional as I am hard pressed to come up with a common use-case for sending zero value invoices on mobile .

emiljoha committed:
Use the same background as similar views.

emiljoha commented:
![any-amount-invoice-payment](https://user-images.githubusercontent.com/12907343/73122054-66b93900-3f80-11ea-8c36-0948f78dd6a7.gif)

emiljoha committed:
Attempt at more polished UI.

emiljoha committed:
Fix issue where writing 500 yeild 0500.

emiljoha committed:
Documentation for checkAmountSuppliedAndGoPayLightningConfirm.

bolatovumar commented on src/action/payment.js:
```diff
@@ -165,14 +165,35 @@ class PaymentAction {
-      this._nav.goPayLightningConfirm();
+      if (
+        this._store.payment.amount === '0' ||
+        this._store.payment.amount === 0
```
is there a need for checking for both `0` and `'0'`? I feel like it should always just be a value of one type. If that's not the case then it should be cast to one type first before checking.

emiljoha replied:
Not really, I should :tm: be '0' its just me being defensive. Will fix.

emiljoha replied:
Fixed in d1c9dfe8

bolatovumar commented on jsconfig.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "compilerOptions": {
```
I'm not sure this file really needs to be in this PR. It's useful but has nothing to do with this particular feature. I think this should be added in a follow-up PR.

emiljoha replied:
Not sure why this is there or what it does... Probably got created when I was trying to set up the project/IDE. Will remove from this PR. Not really sure what it does but I can add it in a follow up PR if It would be useful.

emiljoha replied:
Fixed in 86f0fa3e
Follow up PR: #1331

bolatovumar replied:
@emiljoha it's used by VSCode to configure the built-in JS language service. See here for details: https://code.visualstudio.com/docs/languages/jsconfig

emiljoha replied:
Ah! That makes a lot of sense. Thanks!

bolatovumar commented on src/action/payment.js:
```diff
@@ -199,6 +220,27 @@ class PaymentAction {
+   * @return {Promise<undefined>}
+   */
+  async reEstimateLightningFee() {
+    try {
```
I wonder if the error from this try/catch block should be propagated further from this method so that the UI can display some useful info to the use and try to recover rather than just having a log message.

emiljoha replied:
Thinking about it, that exception will only be thrown if we do not have a valid lightning invoice in the address field. That is dependent on other functions being called before this one. That kind of subtle and implicit dependencies are dangerous, IMHO. I have tried to make the expectations of the method more clear both in code and documentation. 

PS: I am used to writing functional Typescript code in my day job, so all this state and no typing is making me a bit uncomfortable.

emiljoha replied:
Fixed in 93227257 and 51413f29

bolatovumar commented on src/action/payment.js:
```diff
@@ -321,7 +363,8 @@ class PaymentAction {
 
   /**
-   * Send the amount specified in the invoice as a lightning transaction and
+   * Send the amount specified in the payment.amount. If zero use the amount
```
Wonky description here: `If zero use the amount specified in the invoice as a lightning transaction and display the wait screen while the payment confirms.` doesn't really make sense.

emiljoha replied:
Will try to make it more clear.

emiljoha replied:
Fixed in 55985e7d

emiljoha committed:
Remove jsconfig commited by mistake.

emiljoha committed:
Remove unnecesary 0 guard.

emiljoha committed:
Make subtle dependencies more clear and improve error handling.

emiljoha committed:
Make docstring for payLightning more clear.

emiljoha committed:
Change error handling in checkAmountSuppliedAndGoPayLightningConfirm.

Amount will be an empty string when user have not specified in the view before
pressing "pay".

emiljoha commented:
@bolatovumar Thank you for you comments! Have tried to address the issues.

emiljoha commented:
@Roasbeef  (Pinging you as you where the originator of the issue and the only one that have merged commits into to this repository for the last couple of months.)

The change is IMHO ready and has been reviewed by @bolatovumar. Please advice if there are any further steps needed to be taken in order to move forward in solving this issue.

bolatovumar commented:
I have reviewed the PR and it looks good to me. I have not tested it, however.
